## Regression Model

#### Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor
from sklearn.model_selection import RandomizedSearchCV

from sklearn.metrics import (
    r2_score,
    mean_absolute_error,
    root_mean_squared_error
)

In [2]:
import os
os.chdir(r'C:\Users\AnuragS\OneDrive\cv_project\2 Stage Loan Approval & Validation System\data')
os.getcwd()

'C:\\Users\\AnuragS\\OneDrive\\cv_project\\2 Stage Loan Approval & Validation System\\data'

##### Data Fetch

In [3]:
df = pd.read_csv('regression_dataset.csv')
df.head()

,no_of_dependents,education,self_employed,income_annum,loan_term,cibil_score,total_assets,income_per_dependent,asset_coverage_ratio,wealth_cibil_score,loan_amount
0,2,1,0,9600000,12,778,50700000.0,3.200000e+06,1.695652,3.944460e+10,29900000
1,0,0,1,4100000,8,417,17000000.0,4.100000e+06,1.393443,7.089000e+09,12200000
2,3,1,0,9100000,20,506,57700000.0,2.275000e+06,1.942761,2.919620e+10,29700000
3,3,1,0,8200000,8,467,52700000.0,2.050000e+06,1.716612,2.461090e+10,30700000
4,5,0,1,9800000,20,382,55000000.0,1.633333e+06,2.272727,2.101000e+10,24200000


In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 4269 entries, 0 to 4268
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   no_of_dependents      4269 non-null   int64  
 1   education             4269 non-null   int64  
 2   self_employed         4269 non-null   int64  
 3   income_annum          4269 non-null   int64  
 4   loan_term             4269 non-null   int64  
 5   cibil_score           4269 non-null   int64  
 6   total_assets          4269 non-null   float64
 7   income_per_dependent  4269 non-null   float64
 8   asset_coverage_ratio  4269 non-null   float64
 9   wealth_cibil_score    4269 non-null   float64
 10  loan_amount           4269 non-null   int64  
dtypes: float64(4), int64(7)
memory usage: 367.0 KB


#### Define X & y

In [5]:
X = df.drop('loan_amount',axis=1)
y = df['loan_amount']

print(f'Shape of X: {X.shape}')
print(f'Shape of y: {y.shape}')

Shape of X: (4269, 10)
Shape of y: (4269,)


#### Train - test Split

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.20,random_state=42)

print(X_train.shape)
print(X_test.shape)

(3415, 10)
(854, 10)


#### Linear Regression

In [7]:
linear_reg_pipe = Pipeline([
    ('scaler', StandardScaler()),
    ('model', LinearRegression())
])

linear_reg_pipe.fit(X_train, y_train)

y_pred_lr = linear_reg_pipe.predict(X_test)

r2_score_lr = r2_score(y_test, y_pred_lr)
mae_lr = mean_absolute_error(y_test, y_pred_lr)
rmse_lr = root_mean_squared_error(y_test, y_pred_lr)

print(f'r2 Score: {r2_score_lr}')
print(f'MAE: {mae_lr}')
print(f'RMSE: {rmse_lr}')

r2 Score: 0.9216382146067775
MAE: 1898743.5427578473
RMSE: 2506676.4633520595


#### Random Forest

In [8]:
rf = RandomForestRegressor()

rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)

r2_score_rf = r2_score(y_test, y_pred_rf)
mae_rf = mean_absolute_error(y_test, y_pred_rf)
rmse_rf = root_mean_squared_error(y_test, y_pred_rf)

print(f'r2 Score: {r2_score_rf}')
print(f'MAE: {mae_rf}')
print(f'RMSE: {rmse_rf}')

r2 Score: 0.9957699921640399
MAE: 357729.5081967213
RMSE: 582393.7530627861


#### XGBoost

In [9]:
xgb = XGBRegressor()

xgb.fit(X_train, y_train)

y_pred_xgb = xgb.predict(X_test)

r2_score_xgb = r2_score(y_test, y_pred_xgb)
mae_xgb = mean_absolute_error(y_test, y_pred_xgb)
rmse_xgb = root_mean_squared_error(y_test, y_pred_xgb)

print(f'r2 Score: {r2_score_xgb}')
print(f'MAE: {mae_xgb}')
print(f'RMSE: {rmse_xgb}')

r2 Score: 0.9958779215812683
MAE: 410765.125
RMSE: 574915.8125


#### Model Comparison

In [10]:
model_names = ['Linear Regression','Random Forest','XGBoost']
r2_scores = [r2_score_lr,r2_score_rf,r2_score_xgb]
mae_scores = [mae_lr,mae_rf,mae_xgb]
rmse_scores = [rmse_lr,rmse_rf,rmse_xgb]

results = pd.DataFrame({

    'Model':model_names,
    'R2':r2_scores,  
    'MAE':mae_scores,
    'RMSE':rmse_scores
})

results.sort_values('R2', ascending=False)

,Model,R2,MAE,RMSE
2,XGBoost,0.995878,4.107651e+05,5.749158e+05
1,Random Forest,0.995770,3.577295e+05,5.823938e+05
0,Linear Regression,0.921638,1.898744e+06,2.506676e+06


XGBoost Regressor was selected as the final loan amount prediction model because it achieved the highest R² score (99.59%), indicating that it explained the largest proportion of variation in loan amounts. It also produced the lowest RMSE (₹574,916), demonstrating better performance in minimizing large prediction errors. Although Random Forest achieved a slightly lower MAE, XGBoost provided superior overall predictive accuracy and generalization capability. Therefore, XGBoost was chosen as the final regression model.

#### Hyperparameter Tuning (XGBoost)

In [11]:
xgb_hpt = XGBRegressor(random_state=42)

params = {
    'n_estimators':[100,200,500,1000],
    'max_depth':[3,5,7,9],
    'learning_rate':[0.01,0.05,0.1,0.2],
    'subsample':[0.8,0.9,1.0],
    'colsample_bytree':[0.8,0.9,1.0],
    'min_child_weight':[1,3,5]
}

search = RandomizedSearchCV(
    estimator=xgb_hpt,
    param_distributions=params,
    n_iter=30,
    cv=5,
    scoring='r2',
    n_jobs=-1,
    random_state=42,
    verbose=2
)

search.fit(X_train, y_train)

xgb_hpt_tuned = search.best_estimator_

Fitting 5 folds for each of 30 candidates, totalling 150 fits


In [12]:
y_pred_xgb_tuned = xgb_hpt_tuned.predict(X_test)
tuned_r2 = r2_score(y_test, y_pred_xgb_tuned)
print(tuned_r2)

print("R2_tuned_xgb:", r2_score(y_test,y_pred_xgb_tuned))
print("MAE_tuned_xgb:", mean_absolute_error(y_test,y_pred_xgb_tuned))
print("RMSE_tuned_xgb:", root_mean_squared_error(y_test,y_pred_xgb_tuned))

0.9984164834022522
R2_tuned_xgb: 0.9984164834022522
MAE_tuned_xgb: 248653.0625
RMSE_tuned_xgb: 356332.8125


#### XGBoost (Baseline vs Tuned)

In [13]:
pd.DataFrame({
    'model':['xgb','xgb_tuned'],
    'r2_score':[r2_score_xgb,r2_score(y_test,y_pred_xgb_tuned)],
    'mae':[mae_xgb,mean_absolute_error(y_test,y_pred_xgb_tuned)],
    'rmse':[rmse_xgb,root_mean_squared_error(y_test,y_pred_xgb_tuned)]
})

,model,r2_score,mae,rmse
0,xgb,0.995878,410765.1250,574915.8125
1,xgb_tuned,0.998416,248653.0625,356332.8125


Hyperparameter tuning significantly improved the XGBoost Regressor's performance. The tuned model achieved an R² score of 99.84%, an MAE of ₹248,653, and an RMSE of ₹356,333, outperforming the baseline XGBoost model across all evaluation metrics. Therefore, the tuned XGBoost Regressor was selected as the final model for predicting loan amounts.

In [14]:
import joblib

os.makedirs("../model", exist_ok=True)

joblib.dump(xgb_hpt_tuned,"../model/loan_amount_model.pkl")

['../model/loan_amount_model.pkl']